### ISSUE TITLE
Custom Constrained Service Offering vs Qemu - CPU topology doesn't match maximum vcpu count

### ISSUE THREAD

**Issue by StepBee** [apache/cloudstack#6807]
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
CPU definition for qemu
~~~

##### CLOUDSTACK VERSION

~~~
4.17.0.1
~~~

##### CONFIGURATION

We are running a Cloudstack platform based on Ubuntu hosts and KVM.
We have set the guest CPU model to host-mode
guest.cpu.mode=host-model

Phyiscal host CPU config
~~~
lscpu | egrep "Thread\(s\)|Core\(s\)|Socket\(s\)"
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       2
~~~

##### OS / ENVIRONMENT

Ubuntu 20.04.4
Cloudstack Version 4.17.0.1
QEMU emulator version 4.2.1 (Debian 1:4.2-3ubuntu6.23)

##### SUMMARY

When trying to deploy a VM with a "Custom - Contrained" Service offering and defining more than 2 vCPUs, for example 4, we face the following error message:

`2022-10-05 08:12:12,660 INFO  [c.c.v.VirtualMachineManagerImpl] (Work-Job-Executor-65:ctx-395eb8b5 job-95370/job-95371 ctx-9e930d7d) (logid:45597e1f) Unable to start VM on Host {"id": "46", "name": "
xxxxx", "uuid": "c361ef51-a8ab-4ec2-9a32-e5a7ce7886f5", "type"="Routing"} due to unsupported configuration: CPU topology doesn't match maximum vcpu count`

Maybe it's important to note we are trying to deploy the VM via the UI.
We will try to replicate the issue via API.

Using a fixed service offering, providing exactly the same resources, the same host and storage tags, we are able to deploy the VM without issues.

Looking at the generated XML, we find the following CPU related difference - this example is for a VM with 2 VCPU and 4 GB memory, because i couldnt find a way to predict the xml of a failed VM with 4 vCPUs

Custom - Constrained Service Offering (max 16 vCPUs in this example)
~~~
...
  <currentMemory unit='KiB'>4194304</currentMemory>
  <vcpu placement='static' current='2'>16</vcpu>
  <cputune>
    <shares>800</shares>
  </cputune>
...
~~~
and
~~~
...
    <feature policy='disable' name='mpx'/>
    <feature policy='disable' name='intel-pt'/>
    <numa>
      <cell id='0' cpus='0-15' memory='4194304' unit='KiB'/>
    </numa>
  </cpu>
...
~~~

Compared to the XML of the
Fixed Service Offering:

~~~
...
  <currentMemory unit='KiB'>4194304</currentMemory>
  <vcpu placement='static'>2</vcpu>
  <cputune>
    <shares>800</shares>
  </cputune>
...
~~~
and no numa definition
~~~
...
    <feature policy='disable' name='mpx'/>
    <feature policy='disable' name='intel-pt'/>
  </cpu>
...
~~~

Looking at the qemu commandline, we see the following difference

Custom constrained Service Offering
~~~
... -overcommit mem-lock=off -smp 2,maxcpus=16,sockets=16,cores=1,threads=1 -numa node,nodeid=0,cpus=0-15,mem=4096  -uuid 54f79107-bfbc-4320-a845-ef07f896a1fc ...
~~~
versus the fixed Service Offering
~~~
... -overcommit mem-lock=off -smp 2,sockets=2,cores=1,threads=1 -uuid 2e3aba64-e6d8-4b37-a470-5dd075fcc14f ...
~~~

##### STEPS TO REPRODUCE

As i am not sure yet exactly where the rootcause of this issue is, it's hard to say how exactly reproduce in other environments.

In our environment, we encounter the issue only when deploying a VM with
- Custom constrained Service Offering
- specify more than 2 vCPUs (tested 4,8,16)

We don't encounter the issue when deploying a VM with
- Fixed Service Offering
- any valid number of vCPU is possible (tested with 2,4,8)

##### EXPECTED RESULTS

~~~
I would expect to be able to start a VM with the same resource configuration no matter if deployed via custom constrained or fixed Service offering. The result should be the same.
~~~

##### ACTUAL RESULTS

A VM with custom constrained service offering, defining more than 2 vCPUs does not start with the error message

~~~
Unable to start VM on Host {"id": "46", "name": "
xxxxx", "uuid": "c361ef51-a8ab-4ec2-9a32-e5a7ce7886f5", "type"="Routing"} due to unsupported configuration: CPU topology doesn't match maximum vcpu count
~~~
Deploying a VM with the same set of resources (vCPU and memory) using a fixed service offering works without issues.

**Comment by StepBee** [comment]
An update after further testing - i think the issue came with the update to Cloudstack Version 4.17.0.1

I found out i can create a VM with a custom offering and 2 vCPU
`virsh dumpxml <vmid>` shows:
`  <vcpu placement='static' current='2'>16</vcpu>
  <cputune>
    <shares>800</shares>
  </cputune>`

While the VM is running, i can scale the VM with the custom service offering to 4 vCPU, reboot the VM from the operating system (not cloudstack!) and the OS is seeing the 4 vCPUs and is running fine.
`virsh dumpxml <vmid>` shows:
`  <vcpu placement='static' current='4'>16</vcpu>
  <vcpus>
    <vcpu id='0' enabled='yes' hotpluggable='no' order='1'/>
    <vcpu id='1' enabled='yes' hotpluggable='no' order='2'/>
    <vcpu id='2' enabled='yes' hotpluggable='yes' order='3'/>
    <vcpu id='3' enabled='yes' hotpluggable='yes' order='4'/>
    <vcpu id='4' enabled='no' hotpluggable='yes'/>
    <vcpu id='5' enabled='no' hotpluggable='yes'/>
    <vcpu id='6' enabled='no' hotpluggable='yes'/>
    <vcpu id='7' enabled='no' hotpluggable='yes'/>
    <vcpu id='8' enabled='no' hotpluggable='yes'/>
    <vcpu id='9' enabled='no' hotpluggable='yes'/>
    <vcpu id='10' enabled='no' hotpluggable='yes'/>
    <vcpu id='11' enabled='no' hotpluggable='yes'/>
    <vcpu id='12' enabled='no' hotpluggable='yes'/>
    <vcpu id='13' enabled='no' hotpluggable='yes'/>
    <vcpu id='14' enabled='no' hotpluggable='yes'/>
    <vcpu id='15' enabled='no' hotpluggable='yes'/>
  </vcpus>
  <cputune>
    <shares>1600</shares>
  </cputune>`

When i stop the VM from cloudstack and try to start the VM via cloudstack again, it fails with the mentioned error.

Today i looked at a VM, which was created before the update to 4.17.0.1, using exactly the same custom service offering and using 8 vCPUs:
`virsh dumpxml <other-vmid>` shows
`  <vcpu placement='static'>8</vcpu>
  <cputune>
    <shares>4000</shares>
  </cputune>`

There seems to be a difference in 4.17.0.1 in the way the XML for qemu is generated for VMs based on custom contrained service offerings, mainly listing all vcpus in a <vcpus> tag, which seems to conflict with qemu ?

**Comment by DaanHoogland** [comment]
@StepBee thanks for the report. I know some handling of the xml. I don´t know by head which these are though. 
to sum up, and correct me if i am wrong:
- a vm created before 4.17.0.1 can be stopped but won´t start again in 4.17.01.
- a vm created in 4.17.0.1 stops and starts fine.
- ...

**Comment by StepBee** [comment]
@DaanHoogland i still could be totally wrong with my assumptions and i am missing something different, but thats the difference i see at the moment.

I didn't dare the stop the other VM, which is running on 8 vCPU and try to start the VM again - as it's a production VM.

Your summary is not quite correct.

- It looks like before 4.17.0.1 we could create VMs with custom constrained service offerings with more than 2 vCPUs
- since 4.17.0.1 we can't create and run VMs with custom constrained service offering with more than 2 vCPUs anymore
- since 4.17.0.1 we still can create and start a VM with custom constrained service offering with 2 vCPUs and during runtime dynamically scale the number of vCPUs to above 2, with the OS seeing and using them. But this VM will not start anymore if stopped via cloudstack once, complaining with the described qemu error
- But still since 4.17.0.1 we are able to create and start VMs based on fixed service offerings with more than 2 vCPUs

**Comment by StepBee** [comment]
I did additional tests with a Custom unconstrained service offering.

Custom Unconstrained Service Offering
- VM with 2 vCPU deploys and starts
- VM with 4 vCPUs fails with the issue mentioned above

Again i compared the XMLs (i was able to get the 4 vCPU XML by scaling a running 2 vCPU VM to 4 vCPU and not restarting the VM via cloudstack)

`virtsh dumpxml` of 2 vCPU Custom Unconstrained Service offering (**working**):
~~~
  <vcpu placement='static' current='2'>64</vcpu>
  <cputune>
    <shares>800</shares>
  </cputune>
~~~

`virtsh dumpxml` of 4 vCPU Custom Unconstrained Service offering (**not** working):
~~~
  <vcpu placement='static' current='4'>64</vcpu>
  <vcpus>
    <vcpu id='0' enabled='yes' hotpluggable='no' order='1'/>
    <vcpu id='1' enabled='yes' hotpluggable='no' order='2'/>
    <vcpu id='2' enabled='yes' hotpluggable='yes' order='3'/>
    <vcpu id='3' enabled='yes' hotpluggable='yes' order='4'/>
    <vcpu id='4' enabled='no' hotpluggable='yes'/>
    <vcpu id='5' enabled='no' hotpluggable='yes'/>
    <vcpu id='6' enabled='no' hotpluggable='yes'/>
    <vcpu id='7' enabled='no' hotpluggable='yes'/>
    <vcpu id='8' enabled='no' hotpluggable='yes'/>
    <vcpu id='9' enabled='no' hotpluggable='yes'/>
    <vcpu id='10' enabled='no' hotpluggable='yes'/>
    <vcpu id='11' enabled='no' hotpluggable='yes'/>
    <vcpu id='12' enabled='no' hotpluggable='yes'/>
    <vcpu id='13' enabled='no' hotpluggable='yes'/>
    <vcpu id='14' enabled='no' hotpluggable='yes'/>
    <vcpu id='15' enabled='no' hotpluggable='yes'/>
    <vcpu id='16' enabled='no' hotpluggable='yes'/>
    <vcpu id='17' enabled='no' hotpluggable='yes'/>
    <vcpu id='18' enabled='no' hotpluggable='yes'/>
    <vcpu id='19' enabled='no' hotpluggable='yes'/>
    <vcpu id='20' enabled='no' hotpluggable='yes'/>
    <vcpu id='21' enabled='no' hotpluggable='yes'/>
    <vcpu id='22' enabled='no' hotpluggable='yes'/>
    <vcpu id='23' enabled='no' hotpluggable='yes'/>
    <vcpu id='24' enabled='no' hotpluggable='yes'/>
    <vcpu id='25' enabled='no' hotpluggable='yes'/>
    <vcpu id='26' enabled='no' hotpluggable='yes'/>
    <vcpu id='27' enabled='no' hotpluggable='yes'/>
    <vcpu id='28' enabled='no' hotpluggable='yes'/>
    <vcpu id='29' enabled='no' hotpluggable='yes'/>
    <vcpu id='30' enabled='no' hotpluggable='yes'/>
    <vcpu id='31' enabled='no' hotpluggable='yes'/>
    <vcpu id='32' enabled='no' hotpluggable='yes'/>
    <vcpu id='33' enabled='no' hotpluggable='yes'/>
    <vcpu id='34' enabled='no' hotpluggable='yes'/>
    <vcpu id='35' enabled='no' hotpluggable='yes'/>
    <vcpu id='36' enabled='no' hotpluggable='yes'/>
    <vcpu id='37' enabled='no' hotpluggable='yes'/>
    <vcpu id='38' enabled='no' hotpluggable='yes'/>
    <vcpu id='39' enabled='no' hotpluggable='yes'/>
    <vcpu id='40' enabled='no' hotpluggable='yes'/>
    <vcpu id='41' enabled='no' hotpluggable='yes'/>
    <vcpu id='42' enabled='no' hotpluggable='yes'/>
    <vcpu id='43' enabled='no' hotpluggable='yes'/>
    <vcpu id='44' enabled='no' hotpluggable='yes'/>
    <vcpu id='45' enabled='no' hotpluggable='yes'/>
    <vcpu id='46' enabled='no' hotpluggable='yes'/>
    <vcpu id='47' enabled='no' hotpluggable='yes'/>
    <vcpu id='48' enabled='no' hotpluggable='yes'/>
    <vcpu id='49' enabled='no' hotpluggable='yes'/>
    <vcpu id='50' enabled='no' hotpluggable='yes'/>
    <vcpu id='51' enabled='no' hotpluggable='yes'/>
    <vcpu id='52' enabled='no' hotpluggable='yes'/>
    <vcpu id='53' enabled='no' hotpluggable='yes'/>
    <vcpu id='54' enabled='no' hotpluggable='yes'/>
    <vcpu id='55' enabled='no' hotpluggable='yes'/>
    <vcpu id='56' enabled='no' hotpluggable='yes'/>
    <vcpu id='57' enabled='no' hotpluggable='yes'/>
    <vcpu id='58' enabled='no' hotpluggable='yes'/>
    <vcpu id='59' enabled='no' hotpluggable='yes'/>
    <vcpu id='60' enabled='no' hotpluggable='yes'/>
    <vcpu id='61' enabled='no' hotpluggable='yes'/>
    <vcpu id='62' enabled='no' hotpluggable='yes'/>
    <vcpu id='63' enabled='no' hotpluggable='yes'/>
  </vcpus>
  <cputune>
    <shares>1600</shares>
  </cputune>
~~~

**Comment by StepBee** [comment]
In addition i compared again a custom offering vs a fixed offering and maybe discovered the main difference at a different spot.

`virsh dumpxml` Custom constrained/unconstrained offering 4 vCPUs (`not` working):
~~~
  <vcpu placement='static' current='4'>64</vcpu>
  <vcpus>
    <vcpu id='0' enabled='yes' hotpluggable='no' order='1'/>
    <vcpu id='1' enabled='yes' hotpluggable='no' order='2'/>
    <vcpu id='2' enabled='yes' hotpluggable='yes' order='3'/>
    <vcpu id='3' enabled='yes' hotpluggable='yes' order='4'/>
    <vcpu id='4' enabled='no' hotpluggable='yes'/>
...
    <vcpu id='62' enabled='no' hotpluggable='yes'/>
    <vcpu id='63' enabled='no' hotpluggable='yes'/>
  </vcpus>
  <cputune>
    <shares>1600</shares>
  </cputune>
~~~

and a little more down the same xml:

~~~
  <cpu mode='custom' match='exact' check='full'>
    <model fallback='forbid'>Icelake-Server</model>
    <vendor>Intel</vendor>
    <feature policy='require' name='ss'/>
    <feature policy='require' name='vmx'/>
    <feature policy='require' name='hypervisor'/>
    <feature policy='require' name='tsc_adjust'/>
...
    <feature policy='disable' name='intel-pt'/>
    <numa>
      <cell id='0' cpus='0-63' memory='4194304' unit='KiB'/>
    </numa>
  </cpu>
~~~


Compared with
`virsh dumpxml` fixed offering 4 vCPU (**working**):
~~~
  <vcpu placement='static'>4</vcpu>
  <cputune>
    <shares>1600</shares>
  </cputune>
~~~

and a little more down the same xml:

~~~
  <cpu mode='custom' match='exact' check='full'>
    <model fallback='forbid'>Icelake-Server</model>
    <vendor>Intel</vendor>
    <topology sockets='1' cores='4' threads='1'/>
    <feature policy='require' name='ss'/>
    <feature policy='require' name='vmx'/>
    <feature policy='require' name='hypervisor'/>
    <feature policy='require' name='tsc_adjust'/>
...
    <feature policy='disable' name='intel-pt'/>
  </cpu>
~~~


Notice how the **fixed offering** **is defining** the cpus socket / cores / threads
`    <topology sockets='1' cores='4' threads='1'/>`


The **custom offerings** **does not define** the cpu / socket / thread model.

This would kind of make sense looking at the qemu error message
`... due to unsupported configuration: CPU topology doesn't match maximum vcpu count`

What do you think?

**Comment by DaanHoogland** [comment]
@StepBee , I think you are right, but I do not know how and where to fix this in cloudstack, yet. Thanks for the extensive investigation. It seems strange that a `topology` element would only be created for fixed offerings. If so this is the bug for sure.

**Comment by DaanHoogland** [comment]
> https://github.com/apache/cloudstack/blob/13d22c2f542b3570b190bf2eb2acfffae99ef528/plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java#L4762

BTW and FYI your link above didn´t work. I hgad to manually copy it to a new browser tab ;|

**Comment by StepBee** [comment]
@DaanHoogland thanks for the correction.

I tested the agent.properties setting, unfortunately without any change in results.
What makes it even harder to analyze, that i don't have this issue on our staging environment - i perfectly see the cpu topology in the xml on newly created VMs with custom service offering there and can run vms based on custom offerings and >2 vCPUs.

Production and staging have the same setting for the guest cpu mode in agent.properties:
`guest.cpu.mode=host-model`
They have the same settings for the custom service offering, the same settings for the fixed service offering.
For testing i created a new custom constrained service offering on our production system, but the issue stays.

Whats interesting:
In the working XML of a VM based on custom serviceoffering and >2 vCPUs in the staging environment i don't see any lines like:
(taken from the non-working production system which is missing the cpu topology)
~~~
  <vcpus>
    <vcpu id='0' enabled='yes' hotpluggable='no' order='1'/>
    <vcpu id='1' enabled='yes' hotpluggable='no' order='2'/>
    <vcpu id='2' enabled='yes' hotpluggable='yes' order='3'/>
    <vcpu id='3' enabled='yes' hotpluggable='yes' order='4'/>
    <vcpu id='4' enabled='no' hotpluggable='yes'/>
    <vcpu id='5' enabled='no' hotpluggable='yes'/>
    <vcpu id='6' enabled='no' hotpluggable='yes'/>
    <vcpu id='7' enabled='no' hotpluggable='yes'/>
    <vcpu id='8' enabled='no' hotpluggable='yes'/>
    <vcpu id='9' enabled='no' hotpluggable='yes'/>
    <vcpu id='10' enabled='no' hotpluggable='yes'/>
    <vcpu id='11' enabled='no' hotpluggable='yes'/>
    <vcpu id='12' enabled='no' hotpluggable='yes'/>
    <vcpu id='13' enabled='no' hotpluggable='yes'/>
    <vcpu id='14' enabled='no' hotpluggable='yes'/>
    <vcpu id='15' enabled='no' hotpluggable='yes'/>
  </vcpus>
~~~

and i dont see any lines like
(taken from the non-working production system which is missing the cpu topology)
~~~
    <numa>
      <cell id='0' cpus='0-15' memory='4194304' unit='KiB'/>
    </numa>
~~~
which do appear in the wrong XML in our production environment, where i am missing the cpu topology setting.

Maybe thats giving a hint which wrong path the xml build is going in our production (defining `<vcpu id=` lines and defining `<numa>` lines instead of cpu topology setting)?

**Comment by StepBee** [comment]
Just found the cause:
enable.dynamic.scale.vm = true

In our staging environment, this setting is set to "false".
I remember we activated this setting somewhere around the time when we did the update.

Whats **working**:
```
enable.dynamic.scale.vm = **true**
and having a **custom** serviceoffering with **disabled** dynamic scaling
```
or
```
enable.dynamic.scale.vm = **false**
and having a **custom** serviceoffering with **enabled** dynamic scaling (because it has no effect)
```
or
```
enable.dynamic.scale.vm = **true**
and having a **fixed** serviceoffering with **enabled** dynamic scaling
```
or
`using a VM template with disabled dynamic scaling`

Whats **not working**
```
enable.dynamic.scale.vm = **true**
and having a serviceoffering with **enabled** dynamic scaling
and having a vm template with ***enabled** dynamic scaling
```

So enabled dynamical scaling (in global settings AND service offering AND vm template) on **KVM** is responsible for the missing cpu topology line in the qemu xml, which will then result in qemu not starting the VM.

Now the question: Is this how it is supposed to be or is this a bug?

**Comment by DaanHoogland** [comment]
@StepBee , first i am assuming you have the same qemu/libvirt versions in test and production. Which are those?
Not that I think it is in that. Given your description it sounds like a bug to me. I have no idea yet, what it is. It sound like the template being dynamic scaling enabled makes for some processing being skipped.

**Comment by StepBee** [comment]
@DaanHoogland just to avoid confusions, the issue with enabled dynamic scaling is the same for production and testing.

During the process of troubleshooting, i just discovered that i didn't see the issue in my test environment, because dynamic scaling was disabled there.

We are using
Ubuntu 20.04.4
Cloudstack Version 4.17.0.1
QEMU emulator version 4.2.1 (Debian 1:4.2-3ubuntu6.23)

**Comment by DaanHoogland** [comment]
@StepBee , i finally got around trying this in a nested lab env:
```
  <cpu mode='custom' match='exact' check='full'>
    <model fallback='forbid'>qemu64</model>
    <topology sockets='1' cores='4' threads='1'/>
    <feature policy='require' name='x2apic'/>
    <feature policy='require' name='hypervisor'/>
    <feature policy='require' name='lahf_lm'/>
  </cpu>
```
I am on a 4.17.0.0 with ubuntu 20, I'll try to tinker the versions a little to get to reproduce your issue , but so far no cigar.

**Comment by StepBee** [comment]
@DaanHoogland thanks for your efforts.

To be sure, in my setups i am using

agent.properties
`guest.cpu.mode=host-model`

The issue only occurs if dynamic scaling is really enabled meaning you can actually scale a VM while it's running - the settings i am referring to are:
- enable.dynamic.scale.vm = true is set in global settings or
- enable.dynamic.scale.vm = true is set in zone settings
- the service offering supports dynamical scaling
- the vm template supports dynamic scaling

In addition the following conditions have to be met
- the service offering is a custom offering (it doesn't matter if constrained or unconstrained) supporting dynamic scaling
- the number of vCPUs have to be >2

Maybe one of the conditions was missing in your lab env so far?

**Comment by DaanHoogland** [comment]
> * enable.dynamic.scale.vm = true is set in global settings or
> * enable.dynamic.scale.vm = true is set in zone settings

this part I forgot but after doing so I could still deploy and scale a VM (note that my virtual hosts only support 3 cpu (lab restrictions)

id did a `systemctl restart cloudstack-agent` after adding the `guest.cpu.mode=host-model` to the properties file and am now going to do a full reboot to see if that in some way matters

I.E. no successful reproduction of your issue yet

**Comment by DaanHoogland** [comment]
reboots didn´t help either

**Comment by StepBee** [comment]
@DaanHoogland sorry for the confusion - i was under the impression the issue starts >2 vCPUs.
Following your tests i tested exactly 3 vCPU again and the issue does indeed not appear.

I tested with exactly 4 vCPU and the issue happens.

Therefore, i have to correct the conditions -  the issue appears with >3 vCPU, so starting from 4 vCPU

**Comment by StepBee** [comment]
@DaanHoogland i did more tests and the results will very likely help to track down the exact issue in the code
3 vCPUs -> ok
4 vCPUs -> failed
5 vCPUs -> ok
6 vCPUs - failed
7 vCPUs -> ok

I think it's some math operation which is wrong, maybe around this line:
[https://github.com/apache/cloudstack/blob/13d22c2f542b3570b190bf2eb2acfffae99ef528/plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java#L4776](https://github.com/apache/cloudstack/blob/13d22c2f542b3570b190bf2eb2acfffae99ef528/plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java#L4776)

**Comment by DaanHoogland** [comment]
ah, that is a good addition. I am working to get more capacity in my virtual lab. I'll update over the weekend?

**Comment by DaanHoogland** [comment]
I am some closer to reproducing: I have now 8 and my vm with 6 cpu does not have a topology element, but it does run with no problems.

**Comment by DaanHoogland** [comment]
@Stepbee, does the hardware have more than 4 cpu? This could be a problem, and I can´t find any message about it in this issue. (maybe a bit silly)

**Comment by StepBee** [comment]
@DaanHoogland great you could already reach 50% of reproducing, appreciate your efforts.

I can reproduce the issue (topology is missing AND VM does not start) on all different servers i have available in test and production:

~~~
lscpu | egrep "Thread\(s\)|Core\(s\)|Socket\(s\)"
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       2
~~~

~~~
lscpu | egrep "Thread\(s\)|Core\(s\)|Socket\(s\)"
Thread(s) per core:              2
Core(s) per socket:              22
Socket(s):                       2
~~~

~~~
lscpu | egrep "Thread\(s\)|Core\(s\)|Socket\(s\)"
Thread(s) per core:              2
Core(s) per socket:              6
Socket(s):                       2
~~~

If i understood correct you are able to reproduce the issue, that for even numbers the topology element is missing and for odd numbers it's existing, right?

I checked again, and i have the same qemu version in test and production
`QEMU emulator version 4.2.1 (Debian 1:4.2-3ubuntu6.23)`
wich is shipped with Ubuntu 20.04 LTS

Your Version of qemu is probably different?

**Comment by DaanHoogland** [comment]
it reports 4.2 (not 4.2.1) but it is stock ubuntu 20.04.
my gues now is that the `Socket(s):` being 2 could stop it from running correctly.
I have seen the `topology` element being created with 3 and not with 4 cpus requested. but as said my topology is still different:
```
# lscpu | egrep "Thread\(s\)|Core\(s\)|Socket\(s\)"
Thread(s) per core:              1
Core(s) per socket:              1
Socket(s):                       8
```
note that I have only virtual (vmware/cloudstack) based hypervisors.

**Comment by StepBee** [comment]
Which would make sense, seeing 8 sockets in your test setup probably qemu does not require the additional topology information as it can simply assume each vcpu can be a different socket.

**Comment by DaanHoogland** [comment]
finally got a proper reproduction: 8 cores devided over two sockets and the VM won't start!

**Comment by StepBee** [comment]
@DaanHoogland thats great to read!
Finally it looks like we found the exact environment parameters and condition for the issue to happen.

**Comment by DaanHoogland** [comment]
yes, @StepBee good start indeed. I don´t know when I'll have time to address it, but it is on my list for sure. I wish I could say it was fixed.

**Comment by StepBee** [comment]
@DaanHoogland no worries, i highly appreciate your efforts on this.

As a workaround i disabled dynamic scaling in our environments, which was not heavily used so far.
I will try to think through the math operation i was linking as well.

If i can help with any addition information or tests, i will be happy to support.

**Comment by DaanHoogland** [comment]
@StepBee did you see #6870? can you test it?

**Comment by StepBee** [comment]
@DaanHoogland indeed i happily saw it and would we ready to test.

I have to admit i would require a hint how to obtain this patch (i see only 1 file was changed and would require replacement) or the patched release - never tested before and relied on the releases.
Do you have a hint/link for a hint for me?

**Comment by StepBee** [comment]
@DaanHoogland thanks for the hint - i didn't see the most obvious possibility ;-)
The documentation is quite clear, so i'll give the build a go in the next days

**Comment by StepBee** [comment]
I can confirm the patch is working.

Many thanks for your efforts on this one!

Finally i had to dig into building cloudstack from source - a valuable learning ;-)

**Comment by DaanHoogland** [comment]
solved in #6870

**Comment by DaanHoogland** [comment]
solved in #6870
